src/parser/internal/data_type.rs

```diff
@@ -176,19 +177,83 @@ fn optional_simple_data_type(state: &mut State) -> ParseResult<Option<Type>> {
                 b"false" => Ok(Some(Type::False)),
                 b"array" => Ok(Some(Type::Array)),
                 b"callable" => Ok(Some(Type::Callable)),
-                _ => Ok(Some(Type::Identifier(Identifier {
-                    start,
-                    name: id,
-                    end,
-                }))),
+                _ => {
+                    let generics = if state.current.kind == TokenKind::LessThan {
+                        state.next();
+                        let mut generics = vec![];
+                        while state.current.kind != TokenKind::GreaterThan {
+                            let r#type = data_type(state)?;
+                            generics.push(Generic { r#type });
+
+                            if state.current.kind == TokenKind::Comma {
+                                state.next();
+                            } else {
+                                break;
+                            }
+                        }
+
+                        if state.ignored_shift {
+                            utils::skip(state, TokenKind::RightShift)?;
+                            state.ignored_shift = false;
+                        } else if state.current.kind == TokenKind::RightShift {
+                            state.ignored_shift = true;
+                        } else {
+                            utils::skip(state, TokenKind::GreaterThan)?;
+                        }
```

Comment:
this is a "hack", but it works quite well.

`>>` in `Foo<Bar<S>>` gets tokenized as RightShift, so if we see it ( parsing `Bar<S>>` ), set `ignored_shift` on the state to true, and ignore it, once we come back to `Foo<..>>`, skip `>>` if `ignored_shift` is true, and set it to false.

Reply:
In this example:

```
function foo(Bar<S>> $s): void {}
```

since we didn't skip `>>`, the parameters parser will error for "unexpected token `>>` expecting a variable", which is good ( since it errors ), but we could definitely improve it, since we have access to the state, we can check if `ignored_shift`is true, and return an error like "unexpected `>>` token, did you mean `>`?".

Reply:
update: now we error correctly for the code above https://github.com/php-rust-tools/parser/pull/182/commits/d824245235b396427d81a7e2a89843c3012e164c

:)